sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -381,8 +381,25 @@ pi_result cuda_piEventRetain(pi_event event);
 
 /// \endcond
 
+void _pi_queue::compute_stream_wait_for_barrier_if_needed(CUstream stream,
+                                                          pi_uint32 stream_i) {
+  if (barrier_event_ && !compute_applied_barrier_[stream_i]) {
```

Comment:
You are proposing an additional CUDA API call for potentially every operation that might let us eliminate additional calls in the future. Whether this makes sense to do or not depends on whether we expect most streams will have work enqueued to them before or after all the work before the barrier is finished. If the answer is before, the current code will be more performant. If the answer is after, we want to do what you suggest.

My gut feeling says current implementation will be better for most use cases. However the reasoning from the previous paragraph assumes that `cuStreamWaitEvent` and `cuEventQuery` take roughly the same time. Looking into that could give us more information that could inform this decision. However, for now I would leave this as it is.

Reply:
I agree, this is potentially more work up-front, though I would expect `cuEventQuery` to be somewhat lightweight, but doing some benchmarking for it may make sense before a final decision is made on this. I am okay to keep as-is. 😄